controller/internal/enforcer/envoyproxy/authz/server.go

```diff
@@ -0,0 +1,501 @@
+
+	// get the PU context
+	pctxRaw, err := s.puContexts.Get(s.puID)
+	if err != nil {
+		zap.L().Error("ext_authz ingress: failed to get PU context", zap.String("puID", s.puID), zap.Error(err))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "failed to get PU context"), nil
+	}
+	pctx, ok := pctxRaw.(*pucontext.PUContext)
+	if !ok {
+		zap.L().Error("ext_authz ingress: PU context has the wrong type", zap.String("puID", s.puID), zap.String("puContextType", fmt.Sprintf("%T", pctxRaw)))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "PU context has the wrong type"), nil
+	}
+
+	// extract our headers
+	var aporetoKey, aporetoAuth, sourceAdress string
+	var source, dest *ext_authz_v2.AttributeContext_Peer
+	var httpReq *ext_authz_v2.AttributeContext_HttpRequest
+	attrs := checkRequest.GetAttributes()
+	if attrs != nil {
+		source = attrs.GetSource()
+		dest = attrs.GetDestination()
+
+		if source != nil {
+			if addr := source.GetAddress(); addr != nil {
+				if sockAddr := addr.GetSocketAddress(); sockAddr != nil {
+					sourceAdress = sockAddr.GetAddress()
+				}
+			}
+		}
+
+		if request := attrs.GetRequest(); request != nil {
+			httpReq = request.GetHttp()
+			if httpReq != nil {
+				httpReqHeaders := httpReq.GetHeaders()
+				if httpReqHeaders != nil {
+					aporetoKey, _ = httpReqHeaders[aporetoKeyHeader]
+					aporetoAuth, _ = httpReqHeaders[aporetoAuthHeader]
+				}
+			}
+		}
+	}
+
+	flow := newIngressFlowRecord(s.puID, pctx, source, dest, httpReq)
+	defer func() {
+		zap.L().Debug("ext_authz ingress: collecting flow event", zap.String("puID", s.puID), zap.Any("flow", flow))
+		s.collector.CollectFlowEvent(flow)
+	}()
+
+	// if the request is lacking our stuff, we send a 400 back
+	if aporetoAuth == "" || aporetoKey == "" {
+		zap.L().Warn("ext_authz ingress: request missing Aporeto HTTP headers: x-aporeto-key and/or x-aporeto-auth", zap.String("puID", s.puID), zap.String("sourceAddress", sourceAdress))
+		flow.DropReason = "missing Aporeto HTTP headers"
+		return createDeniedCheckResponse(rpc.INVALID_ARGUMENT, envoy_type.StatusCode_BadRequest, "missing headers X-APORETO-KEY or X-APORETO-AUTH"), nil
+	}
+
+	// now we can see if we can decode our claims
+	srcid, scopes, profile, err := s.verifier.ParseToken(aporetoAuth, aporetoKey)
+	if err != nil {
+		zap.L().Debug("ext_authz ingress: failed to parse Aporeto token", zap.String("puID", s.puID), zap.Error(err))
+		flow.DropReason = "failed to parse Aporeto token"
+		return createDeniedCheckResponse(rpc.PERMISSION_DENIED, envoy_type.StatusCode_Forbidden, "failed to parse Aporeto token"), nil
+	}
+	zap.L().Debug("ext_authz ingress: parsed Aporeto token", zap.String("puID", s.puID), zap.String("srcid", srcid), zap.Strings("scopes", scopes), zap.Strings("profile", profile))
+	flow.Source.ID = srcid
+
```

Comment:
I guess there is no API authorization yet .. To be added. I would do the network checks before the API authorization logic. No need to spend time decoding tokens if the network policy rejects it. I think these are backwards.